runners/flink/1.15/src/main/java/org/apache/beam/runners/flink/transform/sql/FlinkSql.java

```diff
@@ -0,0 +1,463 @@
+   */
+  public static <InputT, OutputT> SingleOutputSqlTransform<InputT, OutputT> of(
+      Class<InputT> inputClass, Class<OutputT> outputClass) {
+    return new SingleOutputSqlTransform<>(new FlinkSql<>(inputClass, outputClass));
+  }
+
+  // --------------------- setters ----------------------------
+  /**
+   * Use DDL to define Tables. The DDL string can contain multiple {@code CREATE TABLE} / {@code
+   * CREATE VIEW} statements. The DDL string should not contain any DQL / DML statement.
+   *
+   * @param ddl the table definition
+   * @return this {@link FlinkSql} itself.
+   */
+  FlinkSql<InputT, OutputT> withDDL(String ddl) {
+    ddlList.add(ddl);
+    return this;
+  }
+
+  /**
+   * Use DQL to express the query logic. The query should only contain one DQL, i.e. one top level
+   * {@code SELECT} statement. The query result will be registered as a temporary view. The query
+   * statements comes after this query can refer to the result of this query with the specified
+   * result table name.
+   *
+   * @param resultTableName the table name of the query result.
+   * @param query the SQL DQL statement.
+   */
+  FlinkSql<InputT, OutputT> withQuery(String resultTableName, String query) {
+    queries.put(resultTableName, query);
+    return this;
+  }
+
+  /**
+   * Sets the name of the table created from the input {@link PCollection} to this Sql Transform. By
+   * Default the main input is not used because the tables can be defined from the DDLs.
+   *
+   * @param name the name of the table created from the input {@link PCollection}.
+   */
+  FlinkSql<InputT, OutputT> withMainInputTableName(String name) {
+    if (name.equals(UNUSED_MAIN_INPUT_TABLE_NAME)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Cannot set main input table name to %s which is "
+                  + "reserved for unused main input.",
+              name));
+    }
+
+    if (!mainInputTableName.equals(name)) {
+      TableInfo.InputTableInfo<InputT> mainInputTableInfo = getMainInputTableInfo();
+      inputTableInfo.put(name, mainInputTableInfo);
+      inputTableInfo.remove(mainInputTableName);
+      mainInputTableName = name;
+    }
+    return this;
+  }
+
+  /**
+   * Specify the {@link TypeInformation} of the main input {@link PCollection} for this Sql
+   * Transform. The {@link TypeInformation} is used when converting the main input {@link
+   * PCollection} to the main input {@link Table}.
+   *
+   * <p>By default, the {@link TypeInformation} is a POJO TypeInformation based on the input class.
+   * Users can override the {@link TypeInformation} with this method.
+   *
+   * @param typeInfo the {@link TypeInformation} for the main input {@link PCollection} for this Sql
+   *     Transform.
+   * @see StreamTableEnvironment#createTemporaryView(String, DataStream)
+   */
+  FlinkSql<InputT, OutputT> withMainInputTableTypeInformation(TypeInformation<InputT> typeInfo) {
```

Comment:
Interesting. I was thinking since for avro, we will have the coder (oss AvroCoder) defined, which will contain schema, so in theory we can generate the corresponding Flink TypeInfo object. But it requires specific logic for avro (as you mentioned above), so probably only for LinkedIn and not worth the extra effort.